### ISSUE TITLE
batch sizes in encoder input and decoder output

### ISSUE THREAD

**Issue by cgbhat1978** [espnet/espnet#5953]
I am assuming that the first two dimensions of the tensor 'xs' in the function 'forward' (def forward(self, xs, masks):) of the espent/espnet/nets/transformer/encoder.py (line 300) and 'x' in the function 'forward' (def forward(self, tgt, tgt_mask, memory, memory_mask):) of the espent/espnet/nets/transformer/decoder.py (line 222) have to be the same, they are the batch# and time/maxlen. Only difference in the sizes of these tensors are in the 3rd dimension. Is my assumption correct?
I am writing these tensors out to HDF5 files using the functions below for further analysis. Should I expect that the list of the tensors I get in the two HDF5 files will have the same first two dimensions?
Can you please confirm that my code changes at the encoder, decoder and the code used to read the HDF5 files are correct and will serve the purpose of analysis.

At encoder:
```
def forward(self, xs, masks):
        """Encode input sequence.

        Args:
            xs (torch.Tensor): Input tensor (#batch, time, idim).
            masks (torch.Tensor): Mask tensor (#batch, time).

        Returns:
            torch.Tensor: Output tensor (#batch, time, attention_dim).
            torch.Tensor: Mask tensor (#batch, time).

        """
        self.append_tensor_to_file(xs, 'input_tensors.h5')
 

 def append_tensor_to_file(self, tensor, filename):
        """Appends a tensor to a file - chitra for visualization

    	Args:
    	  tensor: The tensor to be appended.
    	  filename: The name of the file to append to.
    	"""
        tensor_cpu = tensor.cpu()
        numpy_array = tensor_cpu.numpy()
        with h5py.File(filename, 'a') as f:
            # Create a new group
            group = f.require_group('tensors')  # Create or open the group
            dataset_name = f'{len(group)}'  # Generate unique dataset name
            print(dataset_name)
            dataset_name = 'tensor_{}'.format(int(dataset_name)+1)
            print(dataset_name)
            group.create_dataset(dataset_name, data=numpy_array)
```

At the decoder:

```
def forward(self, tgt, tgt_mask, memory, memory_mask):
        """Forward decoder.

        Args:
            tgt (torch.Tensor): Input token ids, int64 (#batch, maxlen_out) if
                input_layer == "embed". In the other case, input tensor
                (#batch, maxlen_out, odim).
            tgt_mask (torch.Tensor): Input token mask (#batch, maxlen_out).
                dtype=torch.uint8 in PyTorch 1.2- and dtype=torch.bool in PyTorch 1.2+
                (include 1.2).
            memory (torch.Tensor): Encoded memory, float32 (#batch, maxlen_in, feat).
            memory_mask (torch.Tensor): Encoded memory mask (#batch, maxlen_in).
                dtype=torch.uint8 in PyTorch 1.2- and dtype=torch.bool in PyTorch 1.2+
                (include 1.2).

        Returns:
            torch.Tensor: Decoded token score before softmax (#batch, maxlen_out, odim)
                   if use_output_layer is True. In the other case,final block outputs
                   (#batch, maxlen_out, attention_dim).
            torch.Tensor: Score mask before softmax (#batch, maxlen_out).

        """

        x = self.embed(tgt)
        x, tgt_mask, memory, memory_mask = self.decoders(
            x, tgt_mask, memory, memory_mask
        )
        if self.normalize_before:
            x = self.after_norm(x)
        if self.output_layer is not None:
            x = self.output_layer(x)
        
        self.append_tensor_to_file(x, 'output_tensors.h5')
        

def append_tensor_to_file(self, tensor, filename):
        """Appends a tensor to a file - chitra for visualization

    	Args:
    	  tensor: The tensor to be appended.
    	  filename: The name of the file to append to.
    	"""
        tensor_cpu = tensor.cpu()
        numpy_array = tensor_cpu.detach().numpy()

        with h5py.File(filename, 'a') as f:
            # Create a new group
            group = f.require_group('tensors')  # Create or open the group
            dataset_name = f'{len(group)}'  # Generate unique dataset name
            print(dataset_name)
            dataset_name = 'tensor_{}'.format(int(dataset_name)+1)
            print(dataset_name)
            group.create_dataset(dataset_name, data=numpy_array)
```

Reading the HDF5 files
```
def load_tensors_from_file(filename):
    tensors = []
    with h5py.File(filename, 'r') as f:
        group = f['tensors']
        for dataset_name in group.keys():
            tensor = torch.tensor(group[dataset_name][:])
            #tensor = tensor.to(device)
            tensors.append(tensor)
    return tensors

input_tensors = load_tensors_from_file('input_tensors.h5')
output_tensors = load_tensors_from_file('output_tensors.h5')
```
